pyoomph/ast2ir.py

```diff
@@ -946,44 +940,40 @@ def _create_union_equals_method(self, union_type: UnionType) -> ir.MethodDef:
         functype = FunctionType(argtypes=[union_type, union_type], returntype=BOOL)
         self_var = ir.LocalVariable(union_type)
         other_var = ir.LocalVariable(union_type)
+        self_matches_var = ir.LocalVariable(BOOL)
+        other_matches_var = ir.LocalVariable(BOOL)
+        result_var = ir.LocalVariable(BOOL)
+        converter = _FunctionOrMethodConverter(
+            self, self.variables.copy(), functype.returntype
+        )
 
-        self_switch_cases: Dict[Type, List[ir.Instruction]] = {}
-        for self_type in union_type.type_members:
-            specific_self_var = ir.LocalVariable(self_type)
-
-            other_switch_cases = {}
-            for other_type in union_type.type_members:
-                specific_other_var = ir.LocalVariable(other_type)
-                result_var = ir.LocalVariable(BOOL)
-                other_switch_cases[other_type] = (
-                    [
-                        ir.GetFromUnion(specific_self_var, self_var),
-                        ir.GetFromUnion(specific_other_var, other_var),
-                        ir.IncRef(specific_self_var),
-                        ir.IncRef(specific_other_var),
-                        ir.CallMethod(
-                            specific_self_var,
-                            "equals",
-                            [specific_other_var],
-                            result_var,
-                        ),
-                        ir.Return(result_var),
-                    ]
-                    if self_type == other_type
-                    else [
-                        ir.VarCpy(result_var, self.variables["false"]),
-                        ir.Return(result_var),
-                    ]
-                )
+        for member_type in union_type.type_members:
+            converter.code.append(
+                ir.UnionMemberCheck(self_matches_var, self_var, member_type)
+            )
+            converter.code.append(
+                ir.UnionMemberCheck(other_matches_var, self_var, member_type)
```

Comment:
```suggestion
                ir.UnionMemberCheck(other_matches_var, other_var, member_type)
```

I wonder what kind of test would make this an error